### ISSUE TITLE
affine_transform does all work with float64

### ISSUE THREAD

**Issue by cancan101** [scipy/scipy#8495]
Looking over the code in [`affine_transform`](https://github.com/scipy/scipy/blob/460571f7bd0208d0dd05ad1e32b7cba91f26b328/scipy/ndimage/interpolation.py#L453-L480), it seems that the input is cast to `float64`. I suggest offering a way to work with eg `float32` for improved performance (memory access and flops).

**Comment by adbugger** [comment]
Hi, is this bug ok to work on? I've been looking through the code think I have an idea of how to go about it. Would another argument (such as dtype) with a default of numpy.float64 be a good approach to this?

**Comment by jaimefrio** [comment]
If you want to work with  `float32` all you have to do is do the filtering yourself, e.g.:

```
filtered_image = ndimage.spline_filter(image, output=np.float32)
transformed_image = ndimage.affine_transform(filtered_image, matrix, prefilter=False)
```

Eventually the code is going to call into the lower level C code (`_nd_image.zoom_shift` or `_nd_image.geometric_transform`), and whatever your `input` or `output` dtypes are, the low level calculations are going to happen in `float64`, and changing that would be a Gargantuan task.

I'm wary of adding more arguments to any of these functions. What could be an option is, if both the input and output arrays fulfill certain conditions to be defined, e.g. like both being `float32`, make the filtered array `float32` instead of `float64`?

The other alternative is to document in the docstring that prefiltering creates a `float64` array, but that this can be avoided by doing the filtering yourself and choosing the dtype explicitly?

**Comment by cancan101** [comment]
Is the underlying c code float 64 only? That wasn't clear to me.

**Comment by jaimefrio** [comment]
Yes, all the internal calculations happens in `float64`. But the input is casted in small chunks as required by the computations, so in most cases the memory overhead of this is relatively small.

**Comment by adbugger** [comment]
Alright, so should I change the docstring to make this clear so that this bug can be closed?

**Comment by cancan101** [comment]
Might be nice to expose the type and then let the function handle the pre filtering rather than complicating the client experience.

**Comment by adbugger** [comment]
So if the input and the ouput both are of type float32, then the filtered and offset arrays can also be casted to type float32, leaving all lower C level computations ( _nd_image*) unchanged. The output type is already casted to float32 if specified, so I don't think that part of the code needs to be touched.

Does changing the type of the filtered and offset arrays based on the above conditions seem like a good way forward?

**Comment by jaimefrio** [comment]
I have added a small comment regarding the temporary `float64` array to the docstrings of `ndimage.interpolation` functions in #8528.

**Comment by cancan101** [comment]
@jaimefrio cool!. Is there any value in offering a dtype flag such that the user doesnt have to prefilter? From what @adbugger said above it sounds like the c level might be able to take advantage of float32 data.
